src/librbd/api/Trash.cc

```diff
@@ -171,6 +170,28 @@ int Trash<I>::move(librados::IoCtx &io_ctx, rbd_trash_image_source_t source,
     return r;
   }
 
+  //migration already issues a request to check for watchers
+  if(source != RBD_TRASH_IMAGE_SOURCE_MIGRATION) {
```

Comment:
I wouldn't add this condition. Checking it twice in that case would not hurt I think.

Reply:
The other reason is `Migration::prepare` opens the image and then calls `Trash::move`. But since there's already a watcher, `Trash::move` fails. This combined with the fact that Migration also checks for watchers, made this check a bit redundant.

Reply:
Then probably change this to `if (source == RBD_TRASH_IMAGE_SOURCE_USER)` or pass a parameter to the function call if we want to check for watchers? @dillaman what do you think?

Reply:
Sure, but is there any specific reason for not using `if(source != RBD_TRASH_IMAGE_SOURCE_MIGRATION)` as a condition?

(I left this unmodified for now, in the latest PR)

Reply:
This condition looks to me as a layering violation: trash_move needs to make same assumptions about the caller. With `if (source == RBD_TRASH_IMAGE_SOURCE_USER)` it looks less uglier, because of simple assumption: we check only when trash is called by users and do not bother with internal cases assuming it is up to the caller to do it if necessary.

A clean solution in my opinion is to move this outside of the function. And if we want code reuse add a wrapper. Or at least pass the parameter so it was the caller who  should decide if we want the check.

But this is nitpicking and just my suggestions. I don't mind much either way.

Reply:
I like the idea of just passing a boolean to say "check for watchers".